File: testflow-master/frontend/src/api/requirement.ts
import api from './index'

// 需求文件相关接口
export interface RequirementFile {
  id: number
  project_id: number
  filename: string
  file_path: string
  file_size: number
  file_type: string
  uploaded_by: number
  upload_time: string
  extracted_content?: string | null
  is_extracted: boolean
  extract_error?: string | null
  uploader?: {
    id: number
    username: string
  }
}

export interface RequirementFileContent {
  id: number
  filename: string
  file_type: string
  extracted_content?: string | null
  is_extracted: boolean
  extract_error?: string | null
  requirement_points?: RequirementPoint[]
}

export interface RequirementPoint {
  id: number
  requirement_file_id: number
  content: string
  order_index: number
  status: 'draft' | 'confirmed' | 'processing' | 'completed'
  created_by_ai: boolean
  edited_by_user: boolean
  created_by: number
  updated_by?: number
  created_at: string
  updated_at: string
  creator?: {
    id: number
    username: string
  }
  updater?: {
    id: number
    username: string
  }
}

// 需求文件API
export const requirementApi = {
  // ========== 需求文件管理 ==========

  // 获取项目需求文件列表
  getFileList: (projectId: number): Promise<RequirementFile[]> => {
    return api.get(`/requirements/${projectId}/files`)
  },

  // 下载需求文件
  downloadFile: (projectId: number, fileId: number): Promise<Blob> => {
    return api.get(`/requirements/${projectId}/files/${fileId}/download`, {
      responseType: 'blob'
    })
  },

  // 重命名文件
  renameFile: (projectId: number, fileId: number, newFilename: string): Promise<RequirementFile> => {
    const formData = new FormData()
    formData.append('new_filename', newFilename)
    return api.put(`/requirements/${projectId}/files/${fileId}/rename`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    })
  },

  // 获取文件内容
  getFileContent: (projectId: number, fileId: number): Promise<RequirementFileContent> => {
    return api.get(`/requirements/${projectId}/files/${fileId}/content`)
  },

  // 删除需求文件
  deleteFile: (projectId: number, fileId: number): Promise<{ message: string }> => {
    return api.delete(`/requirements/${projectId}/files/${fileId}`)
  },

  // ========== 按模块管理需求点 ==========

  // 获取模块的需求点列表
  getModuleRequirementPoints: (projectId: number, moduleId: number): Promise<RequirementPoint[]> => {
    return api.get(`/api/projects/${projectId}/modules/${moduleId}/requirement-points`)
  },

  // 创建需求点
  createModuleRequirementPoint: (
    projectId: number,
    moduleId: number,
    data: { content: string; priority?: string }
  ): Promise<RequirementPoint> => {
    return api.post(`/api/projects/${projectId}/modules/${moduleId}/requirement-points`, null, {
      params: data
    })
  },

  // 更新需求点
  updateModuleRequirementPoint: (
    projectId: number,
    moduleId: number,
    pointId: number,
    data: { content: string; priority?: string }
  ): Promise<RequirementPoint> => {
    return api.put(`/api/projects/${projectId}/modules/${moduleId}/requirement-points/${pointId}`, null, {
      params: data
    })
  },

  // 删除需求点
  deleteModuleRequirementPoint: (projectId: number, moduleId: number, pointId: number): Promise<{ message: string }> => {
    return api.delete(`/api/projects/${projectId}/modules/${moduleId}/requirement-points/${pointId}`)
  },

  // ========== 按模块管理测试点 ==========

  // 获取模块的测试点列表（按需求点分组）
  getModuleTestPoints: (projectId: number, moduleId: number): Promise<{
    requirement_points: Array<{
      id: number
      content: string
      priority: string
      created_by_ai: boolean
      test_points: Array<{
        id: number
        content: string
        test_type: string
        priority: string
        created_by_ai: boolean
      }>
    }>
    statistics: {
      total_requirement_points: number
      total_test_points: number
    }
  }> => {
    return api.get(`/api/projects/${projectId}/modules/${moduleId}/test-points`)
  },

  // 创建测试点
  createModuleTestPoint: (
    projectId: number,
    moduleId: number,
    data: { content: string; requirement_point_id: number; test_type?: string; priority?: string }
  ): Promise<any> => {
    return api.post(`/api/projects/${projectId}/modules/${moduleId}/test-points`, null, {
      params: data
    })
  },

  // 批量创建测试点
  batchCreateModuleTestPoints: (
    projectId: number,
    moduleId: number,
    points: Array<{
      content: string
      test_type?: string
      priority?: string
      requirement_point_id?: number
      created_by_ai?: boolean
    }>,
    clearExisting: boolean = false
  ): Promise<{ success: boolean; created_count: number; deleted_count?: number; points: any[] }> => {
    return api.post(`/api/projects/${projectId}/modules/${moduleId}/test-points/batch`, {
      points,
      clear_existing: clearExisting
    })
  },

  // 更新测试点
  updateModuleTestPoint: (
    projectId: number,
    moduleId: number,
    pointId: number,
    data: { content: string; test_type?: string; priority?: string }
  ): Promise<any> => {
    return api.put(`/api/projects/${projectId}/modules/${moduleId}/test-points/${pointId}`, null, {
      params: data
    })
  },

  // 删除测试点
  deleteModuleTestPoint: (projectId: number, moduleId: number, pointId: number): Promise<{ message: string }> => {
    return api.delete(`/api/projects/${projectId}/modules/${moduleId}/test-points/${pointId}`)
  },

  // ========== 按模块管理测试用例 ==========

  // 获取模块的测试用例列表（按测试点分组）
  getModuleTestCases: (projectId: number, moduleId: number): Promise<{
    test_points: Array<{
      id: number
      content: string
      test_type: string
      priority: string
      created_by_ai: boolean
      test_cases: Array<{
        id: number
        test_point_id: number
        module_id: number
        title: string
        description?: string
        preconditions?: string
        test_steps?: Array<{ step: number; action: string; expected: string }>
        expected_result?: string
        test_method?: string
        test_category?: string
        status: string
        created_by_ai: boolean
        edited_by_user: boolean
        created_at: string
        updated_at?: string
      }>
    }>
    total_test_points: number
    total_test_cases: number
  }> => {
    return api.get(`/api/projects/${projectId}/modules/${moduleId}/test-cases`)
  },

  // 创建测试用例
  createModuleTestCase: (
    projectId: number,
    moduleId: number,
    data: {
      title: string
      test_point_id: number
      description?: string
      preconditions?: string
      test_steps?: Array<{ step: number; action: string; expected: string }>
      expected_result?: string
      test_method?: string
      status?: string
      created_by_ai?: boolean
    }
  ): Promise<any> => {
    return api.post(`/api/projects/${projectId}/modules/${moduleId}/test-cases`, data)
  },

  // 更新测试用例
  updateModuleTestCase: (
    projectId: number,
    moduleId: number,
    caseId: number,
    data: {
      title?: string
      description?: string
      preconditions?: string
      test_steps?: Array<{ step: number; action: string; expected: string }>
      expected_result?: string
      test_method?: string
      status?: string
    }
  ): Promise<any> => {
    return api.put(`/api/projects/${projectId}/modules/${moduleId}/test-cases/${caseId}`, data)
  },

  // 删除测试用例
  deleteModuleTestCase: (projectId: number, moduleId: number, caseId: number): Promise<{ message: string }> => {
    return api.delete(`/api/projects/${projectId}/modules/${moduleId}/test-cases/${caseId}`)
  },

  // 批量创建测试用例
  batchCreateModuleTestCases: (
    projectId: number,
    moduleId: number,
    testCases: Array<{
      title: string
      test_point_id: number
      description?: string
      preconditions?: string
      test_steps?: Array<{ step: number; action: string; expected: string }>
      expected_result?: string
      test_method?: string
      created_by_ai?: boolean
    }>,
    clearExisting: boolean = false
  ): Promise<{ success: boolean; created_count: number; deleted_count?: number; test_cases: any[] }> => {
    return api.post(`/api/projects/${projectId}/modules/${moduleId}/test-cases/batch`, {
      test_cases: testCases,
      clear_existing: clearExisting
    })
  }
}

File: testflow-master/backend/app/models/requirement.py
"""
需求相关数据模型
"""
from typing import List, Optional
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, ForeignKey, DateTime, Boolean, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func
import enum

from app.database import Base


class RequirementStatus(str, enum.Enum):
    """需求点状态枚举"""
    DRAFT = "draft"
    CONFIRMED = "confirmed"
    PROCESSING = "processing"
    COMPLETED = "completed"


class RequirementFile(Base):
    """需求文件模型"""
    __tablename__ = "requirement_files"
    
    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("projects.id"), nullable=False)
    module_id: Mapped[Optional[int]] = mapped_column(ForeignKey("modules.id", ondelete="SET NULL"), index=True)
    filename: Mapped[str] = mapped_column(String(255), nullable=False)
    file_path: Mapped[str] = mapped_column(String(500), nullable=False)
    file_size: Mapped[int] = mapped_column(Integer, nullable=False)
    file_type: Mapped[str] = mapped_column(String(50), nullable=False)
    uploaded_by: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    
    # 文件内容提取
    extracted_content: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_extracted: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    extract_error: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    # 图片信息
    has_images: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    image_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    
    # 时间戳
    upload_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    project: Mapped["Project"] = relationship("Project", back_populates="requirement_files")
    module: Mapped[Optional["Module"]] = relationship("Module", back_populates="requirement_files")
    uploader: Mapped["User"] = relationship("User")
    requirement_points: Mapped[List["RequirementPoint"]] = relationship("RequirementPoint", back_populates="requirement_file")
    images: Mapped[List["RequirementImage"]] = relationship("RequirementImage", back_populates="requirement_file", cascade="all, delete-orphan")
    
    def __repr__(self) -> str:
        return f"RequirementFile(id={self.id!r}, filename={self.filename!r}, project_id={self.project_id!r})"


class RequirementPoint(Base):
    """需求点模型（支持新架构）"""
    __tablename__ = "requirement_points"
    
    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    
    # 关联到需求文件（旧架构，兼容性）
    requirement_file_id: Mapped[Optional[int]] = mapped_column(ForeignKey("requirement_files.id", ondelete="SET NULL"), index=True)
    
    # 关联到模块（新架构）
    module_id: Mapped[Optional[int]] = mapped_column(ForeignKey("modules.id", ondelete="SET NULL"), index=True)
    
    content: Mapped[str] = mapped_column(Text, nullable=False)
    priority: Mapped[str] = mapped_column(String(20), default="medium")  # high/medium/low
    source: Mapped[str] = mapped_column(String(20), default="manual")  # ai_generated/manual
    order_num: Mapped[int] = mapped_column("order_index", Integer, default=0, index=True)  # 排序号，数据库列名为order_index
    
    # 状态（保留以兼容旧数据）
    status: Mapped[Optional[RequirementStatus]] = mapped_column(Enum(RequirementStatus), default=RequirementStatus.DRAFT)
    
    # 编辑标记（保留以兼容旧数据）
    created_by_ai: Mapped[bool] = mapped_column(Boolean, default=False)
    edited_by_user: Mapped[bool] = mapped_column(Boolean, default=False)
    
    # 用户信息
    created_by: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    updated_by: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id"))
    
    # 时间戳
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # 关系
    module: Mapped[Optional["Module"]] = relationship("Module", back_populates="requirement_points")
    requirement_file: Mapped[Optional["RequirementFile"]] = relationship("RequirementFile", back_populates="requirement_points")
    creator: Mapped["User"] = relationship("User", foreign_keys=[created_by])
    updater: Mapped[Optional["User"]] = relationship("User", foreign_keys=[updated_by])
    test_points: Mapped[List["TestPoint"]] = relationship(
        "TestPoint", 
        back_populates="requirement_point",
        foreign_keys="[TestPoint.requirement_point_id]",
        cascade="all, delete-orphan"
    )
    
    def __repr__(self) -> str:
        return f"RequirementPoint(id={self.id!r}, content={self.content[:50]!r}...)"
